pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/event/UploadsEventReceiver.java

```diff
@@ -0,0 +1,138 @@
+ *    </tr>
+ *  </thead>
+ *  <tbody>
+ *    <tr>
+ *      <th>OBJECT_FINALIZE</th>
+ *      <th>
+ *          Updates the {@code uploaded_at} column of the matching record in {@code file_uploads} to
+ *          the {@code publishTime} of the message
+ *      </th>
+ *    </tr>
+ *  </tbody>
+ * </table>
+ */
+@Slf4j
+public class UploadsEventReceiver implements MessageReceiver {
+    private ProjectSubscriptionName subscriptionName;
+
+    private static enum Attribute {
+        NOTIFICATION_CONFIG("notificationConfig"),
+        EVENT_TYPE("eventType"),
+        PAYLOAD_FORMAT("payloadFormat"),
+        BUCKET_ID("bucketId"),
+        OBJECT_ID("objectId"),
+        OBJECT_GENERATION("objectGeneration"),
+        EVENT_TIME("eventTime");
+
+        public final String key;
+
+        private Attribute(String key) {
+            this.key = new String(key);
+        }
+    }
+
+    public UploadsEventReceiver(ProjectSubscriptionName subscriptionName) {
+        this.subscriptionName = subscriptionName;
+    }
+
+    public List<EventType> acceptedEventTypes() {
+        return List.of(EventType.OBJECT_FINALIZE);
+    }
+
+    @Override
+    public void receiveMessage(PubsubMessage message, AckReplyConsumer consumer) {
+        final var messageId = message.getMessageId();
+
+        if (!requiredAttributesArePresent(message)) {
+            log.error("{} received a message missing required attributes [subscription:{}, message:{}]",
+                    this.getClass().getSimpleName(),
+                    subscriptionName,
+                    messageId);
+            consumer.ack();
+            return;
+        }
+        
+        final var eventType = EventType.valueOf(message.getAttributesOrThrow(Attribute.EVENT_TYPE.key));
+        assert eventType != null;
+
+        switch (eventType) {
+            case OBJECT_FINALIZE:
+                handleFinalizeNotification(message);
+                consumer.ack();
+                break;
+            default:
+                // Nothing to do, just ack and continue on
+                consumer.ack();
+        }
+    }
+
+    private boolean requiredAttributesArePresent(PubsubMessage message) {
+        return message.containsAttributes(Attribute.BUCKET_ID.key)
+            && message.containsAttributes(Attribute.OBJECT_ID.key)
+            && message.containsAttributes(Attribute.EVENT_TYPE.key);
+    }
+
+    /**
+     * Handles messages for the `OBJECT_FINALIZE` messages from Google Cloud Storage and updates
+     *  any {@code file_upload} records matching the object's {@code blob_name}.
+     * 
+     * @param message the incoming message from Cloud Storage
+     */
+    private void handleFinalizeNotification(PubsubMessage message) {
+        final var storage = StorageOptions.getDefaultInstance().getService();
+        assert storage != null;
+
+        final var blobName = message.getAttributesOrThrow(Attribute.OBJECT_ID.key);
+        final var publishedAt = message.getPublishTime();
```

Comment:
There are multiple values for the time in the message. This is using the publish time of the event, but there is another time available under the EVENT_TIME attribute in the message. The expected difference between these two values is in the range of nanos/millis, but this can be updated to use the more applicable EVENT_TIME field.

Reply:
No longer applicable.